### ISSUE TITLE
Add floating entity cleanup/control

### ISSUE THREAD

**Issue by dodexahedron** [SEServerExtender/EssentialsPlugin#39]
Since floating entity cleanup is being removed from extender, something should be put into essentials to replace it (though much more capable/useful).

Thoughts on what might be good floating entity management schemes?

**Comment by SilentSymphony** [comment]
I've always thought that Per-User floating objects would be really nice to have, But i have no idea if it is even possible in extender

**Comment by dodexahedron** [comment]
It'd be possible, but it would still have to live within the global limit set on the server.

**Comment by SilentSymphony** [comment]
Well, 25 per user would be a total limit of 375-700 based on some popular servers player counts at 15-28 players, Having them all moving at the same time could get bog down the server so if they were to stop if not affected by gravity it would make hand mining much easier, Gravity collectors, Cannons all without disturbing other users hopefuly, I'm not sure how you'd handle them spawning from ships crashing, or miners/etc, 

Anyway its just a thought, I'm curious to see if anyone has any alternate solutions or concepts to manage them, If i have any other idea's ill be sure to post them.

**Comment by dodexahedron** [comment]
Hmm.
Ownership is the trick, in any case.
Since floating objects aren't really owned by anyone, how would I track this?
I could potentially try to get the nearest player and just assume they own it, for purposes of the limit.

For crashes, or for large mining ships/platforms, that could lead to unexpected results
Any ideas?

**Comment by qrokodial** [comment]
don't know much about how SE does things, but don't floating objects still have some sort of unique ID? can't you just map the ID of the floating objects to the associated players? if the player goes over the limit, just delete the first floating object on the player's stack.

**Comment by dodexahedron** [comment]
Yes, they have unique ids, but how do you know who created them in the first place?

**Comment by qrokodial** [comment]
well, this is a wrapper for the SE server, correct? the SE server has to, at some point, know why its creating these floating items? when it's initally creating these items, wouldn't it need to know why? (so for example: I'm creating floating items here because player X just drilled the shit out of this asteroid).

again, taking shots in the dark for game internals in which I'm not fluent in.

**Comment by zrisher** [comment]
You'd have to amend the core functions that create the objects to call events or attach the entities to the player in some sort of entity manager, this is not currently done at all.

**Comment by dodexahedron** [comment]
Right. This is a wrapper, not a code injector.
In order to do something in the middle of a native game function, that function needs to yield execution to SESE. That's only doable through things like events. If floating object creation is something people would want to hook for other things, though, a good case could be made for adding it to the SE code.  Implementing an event and a public delegate in there would be pretty easy and they just might accept that kind of change.

**Comment by qrokodial** [comment]
I think if we can increase the power of the event system by injecting things like this it would increase the importance of the extender tenfold. just think about how custom Minecraft wrappers/injectors added so much more value to online communities. extension/plugin developers for Minecraft could really go crazy with the things they could do.

this would most likely cause a bit more of a headache when it comes to updating the extender, but I think the power and flexibility an overhauled event system could bring to developers is worth it. what are your thoughts, @dodexahedron?

**Comment by dodexahedron** [comment]
I'm all for having all sorts of events in the game that can be hooked into.
It's about the ultimate in extensibility.
It could potentially be a security risk on the client side, though, if the events could be hooked in scripts or via client-side plugins.
